### ISSUE TITLE
Docker stack traces with free(): invalid pointer when building an image

### ISSUE THREAD

**Issue by aaronjwood** [moby/moby#38504]
**Description**

Tried to do a docker build in the current directory, saw some crashes from C code which Go is calling.

**Steps to reproduce the issue:**
1. sudo docker build -t myimage .

**Describe the results you received:**
```
free(): invalid pointer
SIGABRT: abort
PC=0x7fd698d61077 m=0 sigcode=18446744073709551610

goroutine 0 [idle]:
runtime: unknown pc 0x7fd698d61077
stack: frame={sp:0x7ffd487fb930, fp:0x0} stack=[0x7ffd47ffcd50,0x7ffd487fbd80)
00007ffd487fb830:  0000000000000000  00007fd600000000
00007ffd487fb840:  00000000016fc2a0  00007fd698f7376e
00007ffd487fb850:  00000000016f2a40  00000000487fb8a0
00007ffd487fb860:  0000000000000000  0000000000000000
00007ffd487fb870:  00007fd69897ea61  00007fd69898efb8
00007ffd487fb880:  0000000000000000  44474f0f4c34ce00
00007ffd487fb890:  0000000000000004  0000000001705100
00007ffd487fb8a0:  00007ffd487fb948  0000000000000000
00007ffd487fb8b0:  0000000000000000  0000000000000000
00007ffd487fb8c0:  0000000000000000  00007fd698909de8
00007ffd487fb8d0:  0000000000000000  0000000000000000
00007ffd487fb8e0:  0000000000000000  44474f0f4c34ce00
00007ffd487fb8f0:  0000000000000000  0000000000000002
00007ffd487fb900:  00007ffd487fbca0  44474f0f4c34ce00
00007ffd487fb910:  0000000001705100  00007fd698929d39
00007ffd487fb920:  0000000000000200  00007fd69890b58e
00007ffd487fb930: <0000000000000000  00007fd698f73efc
00007ffd487fb940:  0000000000000000  000000000000003f
00007ffd487fb950:  00000000016f2a30  00007fd698fa3ce4
00007ffd487fb960:  00000000016f2a30  00000000016f2a40
00007ffd487fb970:  00000000016fc170  0000000000000000
00007ffd487fb980:  00007fd6990ab200  00007fd69908b887
00007ffd487fb990:  0000000000000200  00000000000000d7
00007ffd487fb9a0:  00000000016fb840  00007fd698fa3ce4
00007ffd487fb9b0:  fffffffe7fffffff  ffffffffffffffff
00007ffd487fb9c0:  ffffffffffffffff  ffffffffffffffff
00007ffd487fb9d0:  ffffffffffffffff  ffffffffffffffff
00007ffd487fb9e0:  ffffffffffffffff  ffffffffffffffff
00007ffd487fb9f0:  ffffffffffffffff  ffffffffffffffff
00007ffd487fba00:  ffffffffffffffff  ffffffffffffffff
00007ffd487fba10:  ffffffffffffffff  ffffffffffffffff
00007ffd487fba20:  ffffffffffffffff  ffffffffffffffff
runtime: unknown pc 0x7fd698d61077
stack: frame={sp:0x7ffd487fb930, fp:0x0} stack=[0x7ffd47ffcd50,0x7ffd487fbd80)
00007ffd487fb830:  0000000000000000  00007fd600000000
00007ffd487fb840:  00000000016fc2a0  00007fd698f7376e
00007ffd487fb850:  00000000016f2a40  00000000487fb8a0
00007ffd487fb860:  0000000000000000  0000000000000000
00007ffd487fb870:  00007fd69897ea61  00007fd69898efb8
00007ffd487fb880:  0000000000000000  44474f0f4c34ce00
00007ffd487fb890:  0000000000000004  0000000001705100
00007ffd487fb8a0:  00007ffd487fb948  0000000000000000
00007ffd487fb8b0:  0000000000000000  0000000000000000
00007ffd487fb8c0:  0000000000000000  00007fd698909de8
00007ffd487fb8d0:  0000000000000000  0000000000000000
00007ffd487fb8e0:  0000000000000000  44474f0f4c34ce00
00007ffd487fb8f0:  0000000000000000  0000000000000002
00007ffd487fb900:  00007ffd487fbca0  44474f0f4c34ce00
00007ffd487fb910:  0000000001705100  00007fd698929d39
00007ffd487fb920:  0000000000000200  00007fd69890b58e
00007ffd487fb930: <0000000000000000  00007fd698f73efc
00007ffd487fb940:  0000000000000000  000000000000003f
00007ffd487fb950:  00000000016f2a30  00007fd698fa3ce4
00007ffd487fb960:  00000000016f2a30  00000000016f2a40
00007ffd487fb970:  00000000016fc170  0000000000000000
00007ffd487fb980:  00007fd6990ab200  00007fd69908b887
00007ffd487fb990:  0000000000000200  00000000000000d7
00007ffd487fb9a0:  00000000016fb840  00007fd698fa3ce4
00007ffd487fb9b0:  fffffffe7fffffff  ffffffffffffffff
00007ffd487fb9c0:  ffffffffffffffff  ffffffffffffffff
00007ffd487fb9d0:  ffffffffffffffff  ffffffffffffffff
00007ffd487fb9e0:  ffffffffffffffff  ffffffffffffffff
00007ffd487fb9f0:  ffffffffffffffff  ffffffffffffffff
00007ffd487fba00:  ffffffffffffffff  ffffffffffffffff
00007ffd487fba10:  ffffffffffffffff  ffffffffffffffff
00007ffd487fba20:  ffffffffffffffff  ffffffffffffffff

goroutine 1 [syscall]:
runtime.cgocall(0x4af510, 0xc420053cf0, 0x4286c4)
        /usr/lib/go-1.10/src/runtime/cgocall.go:128 +0x64 fp=0xc420053cc0 sp=0xc420053c88 pc=0x403884
github.com/docker/docker-credential-helpers/secretservice._Cfunc_free(0x16f0270)
        _cgo_gotypes.go:111 +0x41 fp=0xc420053cf0 sp=0xc420053cc0 pc=0x4ad4f1
github.com/docker/docker-credential-helpers/secretservice.Secretservice.List.func5(0x16f0270)
        /build/golang-github-docker-docker-credential-helpers-ywm0S2/golang-github-docker-docker-credential-helpers-0.6.1/obj-x86_64-linux-gnu/src/github.com/docker/docker-credential-helpers/secretservice/secretservice_linux.go:96 +0x56 fp=0xc420053d28 sp=0xc420053cf0 pc=0x4aef26
github.com/docker/docker-credential-helpers/secretservice.Secretservice.List(0x0, 0x4fc560, 0xc4200442f0)
        /build/golang-github-docker-docker-credential-helpers-ywm0S2/golang-github-docker-docker-credential-helpers-0.6.1/obj-x86_64-linux-gnu/src/github.com/docker/docker-credential-helpers/secretservice/secretservice_linux.go:97 +0x1de fp=0xc420053dc8 sp=0xc420053d28 pc=0x4ae2ee
github.com/docker/docker-credential-helpers/secretservice.(*Secretservice).List(0x795368, 0x20, 0x28, 0xc420053ea8)
        <autogenerated>:1 +0x35 fp=0xc420053df0 sp=0xc420053dc8 pc=0x4af355
github.com/docker/docker-credential-helpers/credentials.List(0x4fc800, 0x795368, 0x4fc5c0, 0xc42000e018, 0x0, 0x10)
        /build/golang-github-docker-docker-credential-helpers-ywm0S2/golang-github-docker-docker-credential-helpers-0.6.1/obj-x86_64-linux-gnu/src/github.com/docker/docker-credential-helpers/credentials/credentials.go:175 +0x3e fp=0xc420053e78 sp=0xc420053df0 pc=0x4acc9e
github.com/docker/docker-credential-helpers/credentials.HandleCommand(0x4fc800, 0x795368, 0x7ffd487fd8be, 0x4, 0x4fc5a0, 0xc42000e010, 0x4fc5c0, 0xc42000e018, 0x40638c, 0xc42007a058)
        /build/golang-github-docker-docker-credential-helpers-ywm0S2/golang-github-docker-docker-credential-helpers-0.6.1/obj-x86_64-linux-gnu/src/github.com/docker/docker-credential-helpers/credentials/credentials.go:77 +0x13f fp=0xc420053ee8 sp=0xc420053e78 pc=0x4abe5f
github.com/docker/docker-credential-helpers/credentials.Serve(0x4fc800, 0x795368)
        /build/golang-github-docker-docker-credential-helpers-ywm0S2/golang-github-docker-docker-credential-helpers-0.6.1/obj-x86_64-linux-gnu/src/github.com/docker/docker-credential-helpers/credentials/credentials.go:58 +0x1b7 fp=0xc420053f68 sp=0xc420053ee8 pc=0x4abce7
main.main()
        /build/golang-github-docker-docker-credential-helpers-ywm0S2/golang-github-docker-docker-credential-helpers-0.6.1/secretservice/cmd/main_linux.go:9 +0x39 fp=0xc420053f88 sp=0xc420053f68 pc=0x4af3d9
runtime.main()
        /usr/lib/go-1.10/src/runtime/proc.go:198 +0x212 fp=0xc420053fe0 sp=0xc420053f88 pc=0x42ab22
runtime.goexit()
        /usr/lib/go-1.10/src/runtime/asm_amd64.s:2361 +0x1 fp=0xc420053fe8 sp=0xc420053fe0 pc=0x451d51

rax    0x0
rbx    0x7ffd487fbba0
rcx    0x7fd698d61077
rdx    0x0
rdi    0x2
rsi    0x7ffd487fb930
rbp    0x7ffd487fbc80
rsp    0x7ffd487fb930
r8     0x0
r9     0x7ffd487fb930
r10    0x8
r11    0x246
r12    0x7ffd487fbba0
r13    0x1000
r14    0x7fd699113000
r15    0x20
rip    0x7fd698d61077
rflags 0x246
cs     0x33
fs     0x0
gs     0x0
...
<rest of expected docker build output is here>
```

**Describe the results you expected:**
Build image without crashing in native code. It does look like the image is successfully built in the end, regardless of the stack traces.

**Additional information you deem important (e.g. issue happens only occasionally):**
Haven't seen this one before. I just installed Docker on my Ubuntu Cosmic server and tried to build an image.

**Output of `docker version`:**

```
Client:
 Version:           18.09.0
 API version:       1.39
 Go version:        go1.10.4
 Git commit:        4d60db4
 Built:             Wed Nov  7 00:49:01 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.0
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.4
  Git commit:       4d60db4
  Built:            Wed Nov  7 00:16:44 2018
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 10
Server Version: 18.09.0
Storage Driver: btrfs
 Build Version: Btrfs v4.7.3
 Library Version: 101
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: c4446665cb9c30056f4998ed953e6d4ff22c7c39
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: fec3683
Security Options:
 apparmor
 seccomp
  Profile: default
 userns
Kernel Version: 4.18.0-13-generic
Operating System: Ubuntu 18.10
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 31.13GiB
Name: asu64
ID: PWO5:PA62:3YTD:6CJ2:KTYF:RRJD:FRBE:EFVH:QSQ5:6YCS:2FV4:TDNT
Docker Root Dir: /var/lib/docker/1000.1000
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Physical machine running 18.10. Installed Docker using the official instructions on the site but instead of relying on `$(lsb_release -cs)` I just used `bionic` since there is no published stable version for Bionic yet.

**Comment by aaronjwood** [comment]
This one was my fault. I was cleaning up some of my btrfs subvolumes and ended up taking out the ones in `/var/lib/docker/1000.1000/btrfs/subvolumes/`, effectively breaking my install. Reinstalling `docker-ce` was all I needed.

**Comment by aaronjwood** [comment]
I take that back, it seems that the issue remains.

**Comment by thaJeztah** [comment]
Thanks for reporting; The credential-helpers are actually part of the Docker CLI, but looking at the stack dump, the error seems to be coming from the secret service credential-helper itself (not the docker cli);

```
/build/golang-github-docker-docker-credential-helpers-ywm0S2/golang-github-docker-docker-credential-helpers-0.6.1/secretservice/cmd/main_linux.go:9 +0x39 fp=0xc420053f88 sp=0xc420053f68 pc=0x4af3d9
```

That part of the code is only in the secretservice credential helper binary; https://github.com/docker/docker-credential-helpers/blob/v0.6.1/secretservice/cmd/main_linux.go#L9

The Docker CLI itself only vendors the client (there's some files from the `secretservice` in there, but not the one from that error https://github.com/docker/cli/tree/v18.09.0/vendor/github.com/docker/docker-credential-helpers/secretservice)

**Comment by thaJeztah** [comment]
Ah; actually; I see a fix was merged for this: https://github.com/docker/docker-credential-helpers/pull/29

But is not yet in the v0.6.1 tag; https://github.com/docker/docker-credential-helpers/compare/v0.6.1...master


@vdemeester @n4ss - could we tag a new release for the credential-helpers repository?

**Comment by frostmourne-fb** [comment]
I resolved issue by running docker without sudo

**Comment by zinderic** [comment]
Just to share an observation. I'm running Docker for Windows. When I run it against the tcp://localhost:2375 exposed daemon and try to build from another system (Linux subsystem for Windows 10, with env variable DOCKER_HOST=tcp://localhost:2375) the problem is reproduced. But when I try to build it directly via powershell there is no problem.

```
PS > docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 12
Server Version: 18.09.2
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9754871865f7fe2f4e74d43e2fc7ccd237edcbce
runc version: 09c8266bf2fcf9519a651b04ae54c967b9ab86ec
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.125-linuxkit
Operating System: Docker for Windows
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 3.837GiB
Name: linuxkit-00155d645513
ID: BNIC:ACSF:I7HO:EGRF:34HU:WVYS:22XC:AC3V:PJCX:POON:D2HH:I67D
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 22
 Goroutines: 47
 System Time: 2019-02-27T07:36:50.45828Z
 EventsListeners: 1
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 172.30.0.0/16
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine

PS >
```

**Comment by thaJeztah** [comment]
A fix for this was included in the v0.6.2 (and up) releases of the docker-credential-helpers package. We don't ship those helpers as separate packages, but Ubuntu ships them as dependency for their docker-compose package. They're tracking that bug in https://bugs.launchpad.net/ubuntu/+source/golang-github-docker-docker-credential-helpers/+bug/1813003, and would be fixed if they updated to v0.6.2 or up but to work around it, you can read this comment (https://github.com/docker/for-linux/issues/563#issuecomment-491294509); uninstall the `golang-docker-credential-helpers` and `docker-compose` packages, and instead install docker-compose using the instructions in https://docs.docker.com/compose/install/

I'm closing this issue, because of the above, but feel free to continue the conversation
